File: java/side exercises/proverb/src/main/java/Proverb.java
class Proverb {

    private final String[] myWords;
    private final int wordCount;
    private String myProverb = "";

    Proverb(String[] words) {
        myWords = words;
        wordCount = myWords.length;
        makeProverb(0);
    }

    String recite() {
        return myProverb;
    }

    private void makeProverb (int currentWord) {
        if (currentWord == wordCount - 1){
            myProverb += String.format("And all for the want of a %s.", myWords[0]);
        } else if (wordCount > 0){
            myProverb += String.format("For want of a %s the %s was lost.\n",myWords[currentWord], myWords[currentWord + 1]);
            makeProverb(currentWord + 1);
        }
    }

}

File: csharp/side exercises/proverb/Proverb.cs
using System.Collections.Generic;
using System.Linq;

public static class Proverb
{
    private static List<string> proverb = new List<string>();
    private static List<string> mySubjects = new List<string>();

    public static string[] Recite(string[] subjects)
    {
        if (subjects.Length == 0) return subjects;

        proverb.Clear();
        mySubjects = subjects.ToList();
        GenerateProverb(subjects.Length);

        return proverb.ToArray();
    }

    private static void GenerateProverb(int count) {
        if (count == 1) {
            proverb.Add ($"And all for the want of a {mySubjects[0]}.");
        } else {
            proverb.Insert(0, $"For want of a {mySubjects[count - 2]} the {mySubjects[count - 1]} was lost."); 
            GenerateProverb(count - 1);
        }        
    }
}